core/drivers/clk/sam/clk-sam9x60-pll.c

```diff
@@ -0,0 +1,512 @@
+						    unsigned long rate,
+						    unsigned long parent_rate,
+						    bool update)
+{
+	unsigned long tmprate = 0;
+	unsigned long remainder = 0;
+	unsigned long nmul = 0;
+	unsigned long nfrac = 0;
+
+	if (rate < frac->core.charac->output[0].min ||
+	    rate > frac->core.charac->output[0].max)
+		return TEE_ERROR_GENERIC;
+
+	/*
+	 * Calculate the multiplier associated with the current
+	 * divider that provide the closest rate to the requested one.
+	 */
+	nmul = rate / parent_rate;
+	tmprate = parent_rate * nmul;
+	remainder = rate - tmprate;
+
+	if (remainder) {
+		nfrac = UDIV_ROUND_NEAREST((uint64_t)remainder * (1 << 22),
+					   parent_rate);
+
+		tmprate += UDIV_ROUND_NEAREST((uint64_t)nfrac * parent_rate,
+					      (1 << 22));
+	}
+
+	/* Check if resulted rate is a valid. */
+	if (tmprate < frac->core.charac->output[0].min ||
+	    tmprate > frac->core.charac->output[0].max)
+		return TEE_ERROR_GENERIC;
+
+	if (update) {
+		frac->mul = nmul - 1;
+		frac->frac = nfrac;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result sam9x60_frac_pll_set_rate_chg(struct clk *hw,
+						unsigned long rate,
+						unsigned long parent_rate)
+{
+	TEE_Result ret = TEE_SUCCESS;
+	struct sam9x60_frac *frac = hw->priv;
+	struct sam9x60_pll_core *core = &frac->core;
+	vaddr_t regmap = core->base;
+
+	ret = sam9x60_frac_pll_compute_mul_frac(frac, rate, parent_rate, true);
+	if (ret == TEE_SUCCESS) {
+		io_write32(regmap + AT91_PMC_PLL_CTRL1,
+			   SHIFT_U32(frac->mul, core->layout->mul_shift) |
+			   SHIFT_U32(frac->frac, core->layout->frac_shift));
+
+		io_clrsetbits32(regmap + AT91_PMC_PLL_UPDT,
+				AT91_PMC_PLL_UPDT_UPDATE |
+				AT91_PMC_PLL_UPDT_ID_MSK,
+				AT91_PMC_PLL_UPDT_UPDATE | core->id);
+
+		io_setbits32(regmap + AT91_PMC_PLL_CTRL0,
+			     AT91_PMC_PLL_CTRL0_ENLOCK |
+			     AT91_PMC_PLL_CTRL0_ENPLL);
+
+		io_clrsetbits32(regmap + AT91_PMC_PLL_UPDT,
+				AT91_PMC_PLL_UPDT_UPDATE |
+				AT91_PMC_PLL_UPDT_ID_MSK,
+				AT91_PMC_PLL_UPDT_UPDATE | core->id);
+
+		if (WAIT_PLL_READY_TIMEOUT(regmap, core->id))
+			EMSG("PLL not ready");
+	}
+
+	return ret;
+}
+
+static const struct clk_ops sam9x60_frac_pll_ops_chg = {
+	.enable = sam9x60_frac_pll_prepare,
+	.disable = sam9x60_frac_pll_unprepare,
+	.get_rate = sam9x60_frac_pll_recalc_rate,
+	.set_rate = sam9x60_frac_pll_set_rate_chg,
+};
+
+static void sam9x60_div_pll_set_div(struct sam9x60_pll_core *core,
+				    uint32_t div,
+				    bool enable)
+{
+	vaddr_t regmap = core->base;
+	uint32_t ena_msk = enable ? core->layout->endiv_mask : 0;
+	uint32_t ena_val = enable ? (1 << core->layout->endiv_shift) : 0;
+
+	io_clrsetbits32(regmap + AT91_PMC_PLL_CTRL0,
+			core->layout->div_mask | ena_msk,
+			SHIFT_U32(div, core->layout->div_shift) | ena_val);
+
+	io_clrsetbits32(regmap + AT91_PMC_PLL_UPDT,
+			AT91_PMC_PLL_UPDT_UPDATE | AT91_PMC_PLL_UPDT_ID_MSK,
+			AT91_PMC_PLL_UPDT_UPDATE | core->id);
+
+	if (WAIT_PLL_READY_TIMEOUT(regmap, core->id))
+		EMSG("PLL not ready");
```

Comment:
return an error to code `sam9x60_div_pll_set()` and `sam9x60_div_pll_set_rate_chg()`?

Reply:
Update the return code.